lib/block-supports/duotone.php

```diff
@@ -341,25 +301,74 @@ function ( $selector ) use ( $duotone_id ) {
 
 	<?php
 
-	$duotone = ob_get_clean();
+	$svg = ob_get_clean();
+
+	// Clean up the whitespace so it can be used in a data uri.
+	$svg = preg_replace( "/(\r|\n|\t)+/", ' ', $svg );
+	$svg = preg_replace( '/> </', '><', $svg );
+	$svg = trim( $svg );
+
+	$data_uri = 'data:image/svg+xml,' . $svg . '#' . $filter_id;
+
+	// All the variables are already escaped above, so we're not calling esc_url() here.
+	return "url('" . $data_uri . "')";
+}
+
+/**
+ * Render out the duotone stylesheet and SVG.
+ *
+ * @param  string $block_content Rendered block content.
+ * @param  array  $block         Block object.
+ * @return string                Filtered block content.
+ */
+function gutenberg_render_duotone_support( $block_content, $block ) {
+	$block_type = WP_Block_Type_Registry::get_instance()->get_registered( $block['blockName'] );
+
+	$duotone_support = false;
+	if ( $block_type && property_exists( $block_type, 'supports' ) ) {
+		$duotone_support = _wp_array_get( $block_type->supports, array( 'color', '__experimentalDuotone' ), false );
+	}
+
+	$has_duotone_attribute = isset( $block['attrs']['style']['color']['duotone'] );
+
+	if (
+		! $duotone_support ||
+		! $has_duotone_attribute
+	) {
+		return $block_content;
+	}
+
+	$duotone_id      = uniqid();
+	$duotone_colors  = $block['attrs']['style']['color']['duotone'];
+	$filter_property = gutenberg_get_duotone_filter_property( $duotone_id, $duotone_colors );
+
+	$filter_id = 'wp-duotone-filter-' . $duotone_id;
+
+	$scope     = '.' . $filter_id;
+	$selectors = explode( ',', $duotone_support );
+	$scoped    = array();
+	foreach ( $selectors as $sel ) {
+		$scoped[] = $scope . ' ' . trim( $sel );
+	}
+	$selector = implode( ', ', $scoped );
+
+	// !important is needed because these styles render before global styles,
+	// and they should be overriding the duotone filters set by global styles.
+	$filter_style = defined( 'SCRIPT_DEBUG' ) && SCRIPT_DEBUG
+		? $selector . " {\n\tfilter: " . $filter_property . " !important;\n}\n"
+		: $selector . '{filter:' . $filter_property . ' !important;}';
```

Comment:
Would it be safe to add specificity by prepending this with `.wp-site-blocks` instead of using `!important`? Can I count on having that as a container?

Reply:
I think you can rely on `.wp-site-blocks`, but i do wonder if this is a valid use of `!important` - since this is a setting on an individual block, I think it should always take precedence over other CSS.

Reply:
There was mention of a similar situation for elements block supports in the [#core-editor](https://wordpress.slack.com/archives/C02QB2JS7/p1631176228390700) Slack today. 

The linked issue is #33437 and there's a fix by removing `!important` in #34689. That issue is caused by the cascade and because sometimes buttons, that we want to style differently, use the `<a>` tag.

The filter property doesn't cascade the same way, so we're _probably_ fine here. But the elements block supports goes to show that just because we're setting an individual block, `!important` still might not be safe to use.

Reply:
Is there a specific need for the duotone filters to be enqueued before the global styles?

Reply:
Yeah, if we can make it work without the `!important` is nice.

The issue with https://github.com/WordPress/gutenberg/issues/33437 was that `.wp-element-* a` targeted any `a` element in a container (group with paragraph with link + button + social link blocks). So the issue was about how unspecific the selector was. Which makes me think: is it possible that the container with the `.wp-duotone-ID` has an image with duotone and another without it? I'm just not familiar with all the blocks this can affect to. I can't think of any out of my mind.

Reply:
More context about where we use `!important`: we use it in the preset classes, such as `.has-{value}-color` to make sure the user selection is respected despite its low specificity. More context https://github.com/WordPress/gutenberg/pull/29533 

In the case of the link color, it was probably too much due to the wide reach of the selector and side-effects. We're reverting it in this case if the current PR seems solid.